Add a page about React 18 & Concurrent React (#3092)
facebook/react-native-website: docs/react-18-and-react-native.md

---
id: react-18-and-react-native
title: React 18 & React Native
---

import Tabs from '@theme/Tabs'; import TabItem from '@theme/TabItem'; import constants from '@site/core/TabsConstants';

> This page describes how to use React 18 with React Native using the New React Native Architecture.

Reviewer: ```suggestion
This page describes how to use React 18 with React Native using the React Native's New Architecture.
```

I think this reads better and it is more aligned with how we usually refer to the New Architecture.

> **tl;dr:** The first version of React Native compatible with React 18 is **0.69.0**. In order to use the new features in React 18 including automatic batching, `startTransition`, and `useDeferredValue`, you must migrate your React Native app to the New Architecture.

## React 18 and the React Native New Architecture

React 18 introduces [several new features](https://reactjs.org/blog/2022/03/29/react-v18.html) including:

- Automatic batching
- New Strict Mode behaviors
- New hooks (`useId`, `useSyncExternalStore`)

It also includes new concurrent features:

- `startTransition`
- `useTransition`
- `useDeferredValue`
- Full Suspense support

> 

Reviewer 2: `Full Suspense support`
Suspense for **Code Splitting** is actually no supported. We should mention this

The concurrent features in React 18 are built on top of the new concurrent rendering engine. Concurrent rendering is a new behind-the-scenes mechanism that enables React to prepare multiple versions of your UI at the same time.

> Previous versions of React Native built on the old architecture **cannot** support concurrent rendering or concurrent features. This is because the old architecture relied on mutating the native trees, which doesn’t allow for React to prepare multiple versions of the trees at the same time.

Reviewer: ```suggestion
Previous versions of React Native built on the old architecture **cannot** support concurrent rendering or concurrent features. This is because the old architecture relied on mutating the native trees, which doesn’t allow for React to prepare multiple versions of the tree at the same time.
```

> Fortunately, the New Architecture was written bottom-up with concurrent rendering in mind and is fully compatible with React 18. This means, in order to upgrade to React 18 in your React Native app, your application needs to be migrated to the New React Native Architecture including Fabric and TurboModules.

Reviewer: ```suggestion
Fortunately, the New Architecture was written bottom-up with concurrent rendering in mind, and is fully compatible with React 18. This means, in order to upgrade to React 18 in your React Native app, your application needs to be migrated to the React Native's New Architecture including Fabric and TurboModules.
```

---
id: react-18-and-react-native
title: React 18 & React Native
---

import Tabs from '@theme/Tabs'; import TabItem from '@theme/TabItem'; import constants from '@site/core/TabsConstants';

This page describes some of the interactions between React and React Native,
and their implications for the New Architecture of React Native.

If you're interested in using React 18 in your React Native app, this page will also provide meaningful context.

:::info
**tl;dr** The first version of React Native compatible with React 18 is **0.69.0**.
Moreover, to use the [new features of React 18 (i.e. Concurrent React)](https://reactjs.org/blog/2022/03/29/react-v18.html),
you must migrate your React Native app to the New Architecture.
:::

## React and React Native Versions

The version alignment between React and React Native relies on two syncronization points:

> 1. The versions in the `package.json` of the new app template. For example [for React Native 0.68.1](https://github.com/facebook/react-native/blob/b5f1b26a8ff4014e45bbb95c5956ba2c2cf286db/template/package.json#L12-L15) the versions are aligned as follows:

Reviewer 3: ```suggestion
1. The versions in the `package.json` of the new app template. For example [for React Native 0.68.1](https://github.com/facebook/react-native/blob/0.68-stable/template/package.json#L12-L15) the versions are aligned as follows:
```

Let's use link to the the source in stable tag, instead of the latest commit.

> 1. The versions in the `package.json` of the new app template. For example [for React Native 0.68.1](https://github.com/facebook/react-native/blob/b5f1b26a8ff4014e45bbb95c5956ba2c2cf286db/template/package.json#L12-L15) the versions are aligned as follows:

Author: Resolved as this got removed

```json
  "dependencies": {
    "react": "17.0.2",
    "react-native": "0.68.1"
  },
```

2. The React renderers **shipped** with React Native inside the [./Libraries/Renderer](https://github.com/facebook/react-native/tree/main/Libraries/Renderer) of React Native.

This practically means that you **can't bump** the version of React in your `package.json` to a later version,
as you will still be using the older renderer from the folder mentioned above. Bumping the react version in your package.json
will lead to unexpected behaviors.

> For the sake of React 18, the first version of React Native compatible with React 18 is **0.69.0**.

Author: Resolved as this got included

Users on React Native 0.68.0 and previous versions won't be able to use React 18.

If you use the `react-native upgrade` command or the React Native Upgrade Helper,

> you'll bump to the correct React version once you upgrade React Native.

Author: Resolved as this got removed

## React 18 and the React Native New Architecture

React 18 introduced [several new features](https://reactjs.org/blog/2022/03/29/react-v18.html) related to concurrency, and new APIs such as `startTransition`/`useTransition`.

> Please note that in order to use those feature, your React Native application **needs to be migrated** to the New Architecture (specifically needs to be rendering on Fabric).

Author: I'd either link to https://reactnative.dev/docs/new-architecture-intro or defer this to once #3029 has been merged

      return reactRootView;
    }

+   @Override
+   protected boolean isConcurrentRootEnabled() {
+     // If you opted-in for the New Architecture, we enable Concurrent Root (i.e. React 18).
+     // More on this on https://reactjs.org/blog/2022/03/29/react-v18.html
+     return BuildConfig.IS_NEW_ARCHITECTURE_ENABLED;
+   }
  }
}
```

</TabItem>

<TabItem value="kotlin">

```diff
class MainActivity : ReactActivity() {

    open class MainActivityDelegate(activity: ReactActivity?, mainComponentName: String?) : ReactActivityDelegate(activity, mainComponentName) {
        override fun createRootView(): ReactRootView = ReactRootView(context).apply {
            // If you opted-in for the New Architecture, we enable the Fabric Renderer.
            setIsFabric(BuildConfig.IS_NEW_ARCHITECTURE_ENABLED)
        }

+       // If you opted-in for the New Architecture, we enable Concurrent Root (i.e. React 18).
+       // More on this on https://reactjs.org/blog/2022/03/29/react-v18.html
+       override fun isConcurrentRootEnabled() = BuildConfig.IS_NEW_ARCHITECTURE_ENABLED
    }
}
```

</TabItem>
</Tabs>

### Enabling Concurrent React on iOS

> On iOS, you'll have access to the `concurrentRootEnabled` method on your `AppDelegate.mm` file.

Reviewer: ```suggestion
On iOS, you'll have access to the `concurrentRootEnabled` method on your `AppDelegate.mm` file. You should change the returned value to `false` (or `NO`) to disable the feature.
```

```objc
/// This method controls whether the `concurrentRoot`feature of React18 is turned on or off.
///
/// @see: https://reactjs.org/blog/2022/03/29/react-v18.html
/// @note: This requires to be rendering on Fabric (i.e. on the New Architecture).
/// @return: `true` if the `concurrentRoot` feture is enabled. Otherwise, it returns `false`.
- (BOOL)concurrentRootEnabled
{
  // Switch this bool to turn on and off the concurrent root
  return true;
}
```

> ### Users on React Native 0.69 not yet migrated to the New Architecture
> 
> Please note that users on React Native 0.69, but still on the Old Architecture won't benefit from the new React 18 features.
> 
> While you can still override the `isConcurrentRootEnabled` method, that value will have no effect on your rendering.
> You will still be using the old React rendered, you won't benefit of any concurrent features, and you won't be able to use the new
> `startTransition`/`useTransition` APIs.

Reviewer: I'll rephrase this with something more assertive, perhaps even replacing the section with a `caution box`

```suggestion
:::caution

Users on the Old Architecture won't benefit from the new React 18 features, despite being on React Native 0.69.

The `isConcurrentRootEnabled` methods are ignored when running on the new architecture.
:::
```

I think that the box will catch more the attention of the user.
What do you think?

Author: How do you feel about the current wording?

```
### Users on React Native 0.69 not yet migrated to the New Architecture

Note: Users on React Native 0.69, but still on the Old Architecture will still use React 17 mode even if React 18 is installed in the `package.json` file.

Overriding the `isConcurrentRootEnabled` method will have no effect on your app.
```
